### PR TITLE
[GCE] [GCP] UrlMap module

### PR DESCRIPTION
**Pull request by supertom**
##### SUMMARY

GCP UrlMap module

##### COMPONENT NAME

GCE/GCP

##### ISSUE TYPE

 - Feature Pull Request

##### DETAILS
This module provides support for UrlMaps on Google Cloud Platform.  UrlMaps allow users to segment requests by hostname and path and direct those requests to Backend Services.

UrlMaps are a powerful and necessary part of HTTP(S) Global Load Balancing on Google Cloud Platform.

UrlMap takes advantage of the python-api so the appropriate infrastructure has been added to module_utils.

More about UrlMaps can be found at:
https://cloud.google.com/compute/docs/load-balancing/http/url-map

UrlMap API:
https://cloud.google.com/compute/docs/reference/latest/

Google Cloud Platform HTTP(S) Cross-Region Load Balancer:
https://cloud.google.com/compute/docs/load-balancing/http/

##### ANSIBLE VERSION

```
ansible 2.4.0 (ans-http-lb 46c8c88156) last updated 2017/05/09 22:10:38 (GMT +000)
  config file =
  configured module search path = [u'/home/supertom/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /home/supertom/virts/ans-http-lb/code/ansible/lib/ansible
  executable location = /home/supertom/virts/ans-http-lb/code/ansible/bin/ansible
  python version = 2.7.9 (default, Jun 29 2016, 13:08:31) [GCC 4.9.2]
```

##### PLAYBOOK EXAMPLE

A managed instance group, backend service, healthcheck and an instance template are required to use this module.

* A HealthCheck is required in the [console](https://console.cloud.google.com/compute/healthChecks) (currently we don't have a module to create healthchecks)

```
- name: Build GCP UrlMap
  hosts: localhost
  vars_files:
    - vars/auth
  vars:
    mig_one: full-urlmap-mig-east
    mig_two: full-urlmap-mig-west
    mig_one_zone: us-east1-c
    mig_two_zone: us-west1-b
    http_port_name: full-urlmap-httpport
    http_port: 80
    bes: full-urlmap-bes
    # NOTE: THIS HEALTHCHECK MUST BE MADE VIA THE CONSOLE
    # https://console.cloud.google.com/compute/healthChecks
    # (it's just a port 80 check)
    bes_healthcheck: full-urlmap-healthcheck
    urlmap: full-urlmap
    targethttpproxy: gfr-targethttproxy
    instance_template: full-urlmap-template
    external_address_name: gfr-address
  tasks:
    - name: Create Instance template
      gce_instance_template:
        service_account_email: "{{ service_account_email }}"
        credentials_file: "{{ credentials_file }}"
        project_id: "{{ project_id }}"
        name: "{{ instance_template }}"
        image: rhel-7-v20170426
        state: "present"
        tags:
          - http-server
        metadata: 
          startup-script: |
            #!/bin/bash
            yum install -y httpd curl
            systemctl start httpd
            zone=$(curl -sv -H "Metadata-flavor: Google" http://metadata/computeMetadata/v1/instance/zone)
            host=$(curl -sv -H "Metadata-flavor: Google" http://metadata/computeMetadata/v1/instance/name)
            echo "${host}-${zone}" > /var/www/html/index.html
    # ============================================================
    # Create MIG one
    # ============================================================
    - name: Create MIG one
      gce_mig:
        service_account_email: "{{ service_account_email }}"
        credentials_file: "{{ credentials_file }}"
        project_id: "{{ project_id }}"
        name: "{{ mig_one }}"
        zone: "{{ mig_one_zone }}"
        state: "present"
        size: 1
        autoscaling:
          enabled: yes
          name: "{{ mig_one }}"
          policy:
            min_instances: 1
            max_instances: 3
            cool_down_period: 60
            cpu_utilization:
              target: 0.6
        template: "{{ instance_template }}"
        named_ports:
          - name: "{{ http_port_name }}"
            port: "{{ http_port }}"
    # ============================================================
    # Create MIG two
    # ============================================================
    - name: Create MIG two
      gce_mig:
        service_account_email: "{{ service_account_email }}"
        credentials_file: "{{ credentials_file }}"
        project_id: "{{ project_id }}"
        name: "{{ mig_two }}"
        zone: "{{ mig_two_zone }}"
        state: "present"
        size: 1
        autoscaling:
          enabled: yes
          name: "{{ mig_two }}"
          policy:
            min_instances: 1
            max_instances: 3
            cool_down_period: 60
            cpu_utilization:
              target: 0.6
        template: "{{ instance_template }}"
        named_ports:
          - name: "{{ http_port_name }}"
            port: "{{ http_port }}"
    # ============================================================
    # Create Backend Service
    # ============================================================
    - name: Create BES
      gcp_backend_service:
        service_account_email: "{{ service_account_email }}"
        credentials_file: "{{ credentials_file }}"
        project_id: "{{ project_id }}"
        backend_service_name: "{{ bes }}"
        backends:
          - instance_group: "{{ mig_one }}"
            balancing_mode: RATE
            max_rate_per_instance: 10.00
          - instance_group: "{{ mig_two }}"
        healthchecks:
          - "{{ bes_healthcheck }}"
        port_name: "{{ http_port_name }}"
        protocol: HTTP
        timeout: 60
        state: "present"
      register: result
    - name: "assert backend service present"
      assert:
        that:
          - 'result.state == "present"'
    # ============================================================
    # Create URLMap
    # ============================================================
    - name: Create URLMap
      gcp_url_map:
        service_account_email: "{{ service_account_email }}"
        credentials_file: "{{ credentials_file }}"
        project_id: "{{ project_id }}"
        url_map_name: "{{ urlmap }}"
        default_service: "{{ bes }}"
        state: "present"
      register: result
    - name: "assert urlmap present"
      assert:
        that:
          - 'result.state == "present"'
```

### CODE REVIEW

**Comment by supertom** [comment]
@ryansb PTAL.  These latest failures seem to be from yum/jinja2 issues.  I believe I've resolved the python3 and pep8 issues

File: lib/ansible/modules/cloud/google/gcp_url_map.py [hunk 1]
```diff
@@ -0,0 +1,502 @@
...
+  path_matchers:
+    description:
+       - The list of named PathMatchers to use against the URL. Contains
+         path_rules, which is a list of paths and an associated service. A
+         default_service can also be specified for each path_matcher.
+    required: false
+'''
+
+EXAMPLES = '''
+- name: Create Minimum Url_Map
```
Comment (ryansb) [inline lib/ansible/modules/cloud/google/gcp_url_map.py]: s/Minimum/Minimal/

File: lib/ansible/modules/cloud/google/gcp_url_map.py [hunk 2]
```diff
@@ -0,0 +1,502 @@
...
+    description:
+       - Default Backend Service if no host rules match.
+    required: true
+  host_rules:
+    description:
+       - The list of HostRules to use against the URL. Contains
+         host_rules, which is a list of hosts and an associated path_matcher.
+    required: false
+  path_matchers:
+    description:
```
Comment (ryansb) [inline lib/ansible/modules/cloud/google/gcp_url_map.py]: For these complex types, feel free to go more verbose about the expected data structure. I see the link at the top for the API (https://cloud.google.com/compute/docs/reference/latest/urlMaps#resource) has really good info on the nested type.

File: lib/ansible/modules/cloud/google/gcp_url_map.py [hunk 3]
```diff
@@ -0,0 +1,502 @@
...
+                msg="Cannot delete unknown url_map: %s" %
+                (params['url_map_name']))
+        else:
+            # Create
+            (changed, json_output['url_map']) = create_url_map(client,
+                                                               params=params,
+                                                               project_id=conn_params['project_id'])
+    elif params['state'] == 'absent':
+        # Delete
+        (changed, json_output['url_map']) = delete_url_map(client,
```
Comment (ryansb) [inline lib/ansible/modules/cloud/google/gcp_url_map.py]: These parens aren't necessary for unpacking the return values.

**Review by ryansb — COMMENTED** [review]
High level question: is there a case where someone might need to *add* a path matcher to an existing URL mapping, but not specify all the existing ones? I'd expect people to set them all in one place, but I can imagine a user saying "I want to append a path matcher to the list." 

I see the update function just takes the arguments provided and updates the URL map instead of checking parameters. Is there anywhere that might get us in trouble?

File: lib/ansible/modules/cloud/google/gcp_url_map.py (hunk 2 above)
Comment (supertom) [reply lib/ansible/modules/cloud/google/gcp_url_map.py]: Done.

File: lib/ansible/modules/cloud/google/gcp_url_map.py (hunk 1 above)
Comment (supertom) [reply lib/ansible/modules/cloud/google/gcp_url_map.py]: Done.

File: lib/ansible/modules/cloud/google/gcp_url_map.py (hunk 3 above)
Comment (supertom) [reply lib/ansible/modules/cloud/google/gcp_url_map.py]: Done.

File: lib/ansible/modules/cloud/google/gcp_url_map.py [hunk 4]
```diff
@@ -0,0 +1,519 @@
...
+       - The 'name' parameter to which this path_matcher is referred by the
+         host_rule.
+       - The 'default_service' parameter is the name of the
+         BackendService resource. This will be used if none of the path_rules
+         defined by this path_matcher is matched by the URL's path portion.
+       - The 'path_rules' parameter is a list of dictionaries containing a
+         list of paths and a service to direct traffic to. Each path item must
+         start with / and the only place a * is allowed is at the end following
+         a /. The string fed to the path matcher does not include any text after
+         the first ? or #, and those chars are not allowed here.
```
Comment (ryansb) [inline lib/ansible/modules/cloud/google/gcp_url_map.py:79]: Thanks! 👍

**Review by ryansb — APPROVED** [review]

**Comment by supertom** [comment]
@ryansb To answer your high-level question.  I think you're eluding to patch semantics, which some of the GCP APIs support (including UrlMap).  In the Ansible modules we've created so far we've opted to skip those in favor of an update, working under the assumption that the playbook was the "source of truth".  

I don't see why it couldn't be added in the future, though, if users would want it.  We just need some way to signal the module to use patch rather than update in those cases (I guess some field, like 'update_type: patch').

**Comment by ryansb** [comment]
I see - that makes sense.

**Pull request merged**